Marlin/src/module/fancheck.cpp

```diff
@@ -0,0 +1,198 @@
+  bool FanCheck::measuring = false;
+#endif
+bool FanCheck::tacho_state[TACHO_COUNT];
+uint16_t FanCheck::edge_counter[TACHO_COUNT];
+uint8_t FanCheck::rps[TACHO_COUNT];
+FanCheck::Error_Code FanCheck::error_code = FanCheck::Error_Code::NONE;
+bool FanCheck::enabled;
+
+void FanCheck::init() {
+  #define _TACHINIT(N) TERN(E##N##_FAN_TACHO_PULLUP, SET_INPUT_PULLUP, TERN(E##N##_FAN_TACHO_PULLDOWN, SET_INPUT_PULLDOWN, SET_INPUT))(E##N##_FAN_TACHO_PIN)
+  #if HAS_E0_FAN_TACHO
+    _TACHINIT(0);
+  #endif
+  #if HAS_E1_FAN_TACHO
+    _TACHINIT(1);
+  #endif
+  #if HAS_E2_FAN_TACHO
+    _TACHINIT(2);
+  #endif
+  #if HAS_E3_FAN_TACHO
+    _TACHINIT(3);
+  #endif
+  #if HAS_E4_FAN_TACHO
+    _TACHINIT(4);
+  #endif
+  #if HAS_E5_FAN_TACHO
+    _TACHINIT(5);
+  #endif
+  #if HAS_E6_FAN_TACHO
+    _TACHINIT(6);
+  #endif
+  #if HAS_E7_FAN_TACHO
+    _TACHINIT(7);
+  #endif
+}
+
+void FanCheck::update_tachometers() {
+  bool status;
+
+  #define _TACHO_CASE(N) case N: status = READ(E##N##_FAN_TACHO_PIN); break;
+  LOOP_L_N(f, TACHO_COUNT) {
+    switch (f) {
+      #if HAS_E0_FAN_TACHO
+        _TACHO_CASE(0)
+      #endif
+      #if HAS_E1_FAN_TACHO
+        _TACHO_CASE(1)
+      #endif
+      #if HAS_E2_FAN_TACHO
+        _TACHO_CASE(2)
+      #endif
+      #if HAS_E3_FAN_TACHO
+        _TACHO_CASE(3)
+      #endif
+      #if HAS_E4_FAN_TACHO
+        _TACHO_CASE(4)
+      #endif
+      #if HAS_E5_FAN_TACHO
+        _TACHO_CASE(5)
+      #endif
+      #if HAS_E6_FAN_TACHO
+        _TACHO_CASE(6)
+      #endif
+      #if HAS_E7_FAN_TACHO
+        _TACHO_CASE(7)
+      #endif
+      default: continue;
+    }
+
+    if (status != tacho_state[f]) {
+      if (measuring) ++edge_counter[f];
+      tacho_state[f] = status;
+    }
+  }
+}
+
+void FanCheck::compute_speed(int elapsedTime) {
+  LOOP_L_N(f, TACHO_COUNT) {
+    switch (f) {
+      TERN_(HAS_E0_FAN_TACHO, case 0:)
+      TERN_(HAS_E1_FAN_TACHO, case 1:)
+      TERN_(HAS_E2_FAN_TACHO, case 2:)
+      TERN_(HAS_E3_FAN_TACHO, case 3:)
+      TERN_(HAS_E4_FAN_TACHO, case 4:)
+      TERN_(HAS_E5_FAN_TACHO, case 5:)
+      TERN_(HAS_E6_FAN_TACHO, case 6:)
+      TERN_(HAS_E7_FAN_TACHO, case 7:)
+        rps[f] = edge_counter[f] * float(250) / elapsedTime;
+        edge_counter[f] = 0;
+        break;
+    }
+  }
+}
+
+void FanCheck::check_speed() {
+  constexpr int8_t max_extruder_fan_errors = 15000 / Temperature::fan_autocheck_interval_ms;
+  static uint8_t errors_count[TACHO_COUNT];
+
+  #if HAS_AUTO_FAN
+    const uint8_t fanState = thermalManager.get_fan_state();
+  #endif
+
+  bool fans_are_ok = true;
+  LOOP_L_N(f, TACHO_COUNT) {
+    if (rps[f] < 20 && TERN1(HAS_AUTO_FAN, TEST(fanState, f))) {
+      fans_are_ok = false;
+      if (enabled && errors_count[f] >= max_extruder_fan_errors && error_code == Error_Code::NONE) {
+        errors_count[f] = 0;
+        // TODO signal fan (f) speed fault
+      }
+      else ++errors_count[f];
```

Comment:
`if (!enabled) ++errors_count[f]` ?

Reply:
I'm following what prusa firmware does. It seems that it send a message to lcd and then it changes error code. This way it raises error just once and it reset errors count to properly detect next fault. Now I'm not sure if I can send some messages to lcd inside temp isr. I suspect I can't

Reply:
Is it safe to send lcd faults while in temperature isr?

Reply:
You can see that `errors_count[f]` will just keep incrementing, eventually overflowing back to 0. Is it actually counting errors, or is it counting up the number of calls to this code while `!enabled` and calling those "errors"?

Reply:
error count overflow is already fixed in my repo PR and enabled is somethig that may be added to LCD to bypass check for any reason user needs